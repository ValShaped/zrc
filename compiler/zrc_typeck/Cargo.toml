[package]
name = "zrc_typeck"
version = "0.1.0"
edition = "2024"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
derive_more = { version = "2.0.1", features = ["display"] }
indexmap = "2.2.6"
zrc_diagnostics = { version = "0.1.0", path = "../zrc_diagnostics" }
zrc_parser = { version = "0.1.0", path = "../zrc_parser" }
zrc_utils = { version = "0.1.0", path = "../zrc_utils" }
