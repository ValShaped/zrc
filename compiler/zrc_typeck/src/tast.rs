//! Type-checked AST for the Zirco programming language
//!
//! This is similar to the [standard AST](zrc_parser::ast), but with type
//! information attached to it.

pub mod expr;
pub mod stmt;
pub mod ty;
