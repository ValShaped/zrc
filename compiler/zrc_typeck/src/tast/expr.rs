//! Expression representation for the Zirco [TAST](super)

use std::fmt::Display;

/// An [expression kind](TypedExprKind) with its yielded [result type](super::ty::Type) attached to it.
#[derive(PartialEq, Debug, Clone)]
#[allow(clippy::module_name_repetitions)]
pub struct TypedExpr(pub super::ty::Type, pub TypedExprKind);

/// The kind of a [`TypedExpr`]
#[derive(PartialEq, Debug, Clone)]
pub enum TypedExprKind {
    /// `a, b`
    Comma(Box<TypedExpr>, Box<TypedExpr>),

    /// `a = b`
    Assignment(Box<TypedExpr>, Box<TypedExpr>),

    /// `!x`
    UnaryNot(Box<TypedExpr>),
    /// `~x`
    UnaryBitwiseNot(Box<TypedExpr>),
    /// `-x`
    UnaryMinus(Box<TypedExpr>),
    /// `&x`
    UnaryAddressOf(Box<TypedExpr>),
    /// `*x`
    UnaryDereference(Box<TypedExpr>),

    /// `a[b]`
    Index(Box<TypedExpr>, Box<TypedExpr>),
    /// `a.b`
    Dot(Box<TypedExpr>, String),
    /// `a->b`
    Arrow(Box<TypedExpr>, String),
    /// `a(b, c, d, ...)`
    Call(Box<TypedExpr>, Vec<TypedExpr>),

    /// `a ? b : c`
    Ternary(Box<TypedExpr>, Box<TypedExpr>, Box<TypedExpr>),

    /// `a && b`
    LogicalAnd(Box<TypedExpr>, Box<TypedExpr>),
    /// `a || b`
    LogicalOr(Box<TypedExpr>, Box<TypedExpr>),

    /// `a == b`
    Equals(Box<TypedExpr>, Box<TypedExpr>),
    /// `a != b`
    NotEquals(Box<TypedExpr>, Box<TypedExpr>),

    /// `a & b`
    BitwiseAnd(Box<TypedExpr>, Box<TypedExpr>),
    /// `a | b`
    BitwiseOr(Box<TypedExpr>, Box<TypedExpr>),
    /// `a ^ b`
    BitwiseXor(Box<TypedExpr>, Box<TypedExpr>),

    /// `a > b`
    GreaterThan(Box<TypedExpr>, Box<TypedExpr>),
    /// `a >= b`
    GreaterThanOrEqualTo(Box<TypedExpr>, Box<TypedExpr>),
    /// `a < b`
    LessThan(Box<TypedExpr>, Box<TypedExpr>),
    /// `a <= b`
    LessThanOrEqualTo(Box<TypedExpr>, Box<TypedExpr>),

    /// `a >> b`
    BitwiseRightShift(Box<TypedExpr>, Box<TypedExpr>),
    /// `a << b`
    BitwiseLeftShift(Box<TypedExpr>, Box<TypedExpr>),

    /// `a + b`
    Addition(Box<TypedExpr>, Box<TypedExpr>),
    /// `a - b`
    Subtraction(Box<TypedExpr>, Box<TypedExpr>),

    /// `a * b`
    Multiplication(Box<TypedExpr>, Box<TypedExpr>),
    /// `a / b`
    Division(Box<TypedExpr>, Box<TypedExpr>),
    /// `a % b`
    Modulo(Box<TypedExpr>, Box<TypedExpr>),

    /// `x as T`
    Cast(Box<TypedExpr>, super::ty::Type),

    /// Any numeric literal.
    NumberLiteral(String),
    /// Any string literal.
    StringLiteral(String),
    /// Any identifier.
    Identifier(String),
    /// Any boolean literal.
    BooleanLiteral(bool),
}

impl Display for TypedExprKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::NumberLiteral(n) => write!(f, "{n}"),
            Self::StringLiteral(s) => write!(f, "{s}"),
            Self::Identifier(i) => write!(f, "{i}"),
            Self::BooleanLiteral(b) => write!(f, "{b}"),
            Self::Assignment(l, r) => write!(f, "{l} = {r}"),
            Self::Addition(l, r) => write!(f, "{l} + {r}"),
            Self::Subtraction(l, r) => write!(f, "{l} - {r}"),
            Self::Multiplication(l, r) => write!(f, "{l} * {r}"),
            Self::Division(l, r) => write!(f, "{l} / {r}"),
            Self::Modulo(l, r) => write!(f, "{l} % {r}"),
            Self::BitwiseAnd(l, r) => write!(f, "{l} & {r}"),
            Self::BitwiseOr(l, r) => write!(f, "{l} | {r}"),
            Self::BitwiseXor(l, r) => write!(f, "{l} ^ {r}"),
            Self::BitwiseLeftShift(l, r) => write!(f, "{l} << {r}"),
            Self::BitwiseRightShift(l, r) => write!(f, "{l} >> {r}"),
            Self::GreaterThan(l, r) => write!(f, "{l} > {r}"),
            Self::GreaterThanOrEqualTo(l, r) => write!(f, "{l} >= {r}"),
            Self::LessThan(l, r) => write!(f, "{l} < {r}"),
            Self::LessThanOrEqualTo(l, r) => write!(f, "{l} <= {r}"),
            Self::Equals(l, r) => write!(f, "{l} == {r}"),
            Self::NotEquals(l, r) => write!(f, "{l} != {r}"),
            Self::LogicalAnd(l, r) => write!(f, "{l} && {r}"),
            Self::LogicalOr(l, r) => write!(f, "{l} || {r}"),
            Self::Comma(l, r) => write!(f, "{l}, {r}"),
            Self::UnaryNot(e) => write!(f, "!{e}"),
            Self::Cast(x, t) => write!(f, "{x} as {t}"),
            Self::UnaryBitwiseNot(e) => write!(f, "~{e}"),
            Self::UnaryMinus(e) => write!(f, "-{e}"),
            Self::UnaryAddressOf(e) => write!(f, "&{e}"),
            Self::UnaryDereference(e) => write!(f, "*{e}"),
            Self::Arrow(l, r) => write!(f, "{l}->{r}"),
            Self::Ternary(a, b, c) => write!(f, "{a} ? {b} : {c}"),
            Self::Index(a, b) => write!(f, "{a}[{b}]"),
            Self::Dot(a, b) => write!(f, "{a}.{b}"),
            Self::Call(a, b) => write!(
                f,
                "{a}({})",
                b.iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
        }
    }
}

impl Display for TypedExpr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "(({}) as ({}))", self.1, self.0)
    }
}
