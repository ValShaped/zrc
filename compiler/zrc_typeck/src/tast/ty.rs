//! Type representation for the Zirco [TAST](super)
//!
//! Note: In the TAST, we do not assign types a span, because
//! they may be obtained from an arbitrary location in the program
//! (and simply inferred). If it exists *in* the source file, it is
//! given an explicit span (e.g. argument types) however.

use std::fmt::Display;

use derive_more::Display;
use indexmap::IndexMap;

use super::stmt::ArgumentDeclarationList;

/// Data attached to a [`Type::Fn`]
#[derive(Debug, Clone, PartialEq, Display)]
#[display("(fn({arguments}) -> {returns})")]
pub struct Fn<'input> {
    /// The function's arguments
    pub arguments: ArgumentDeclarationList<'input>,
    /// The function's return type
    pub returns: Box<Type<'input>>,
}

/// Auxillary data attached to a [`Fn`] in the
/// [`crate::typeck::scope::GlobalScope`]
#[derive(Debug)]
pub struct FunctionDeclarationGlobalMetadata<'input> {
    /// The corresponding [`Fn`] we wrap
    pub fn_type: Fn<'input>,
    /// If a declaration exists to implement this function
    /// (Only one may exist)
    pub has_implementation: bool,
}

/// The possible Zirco types
#[derive(PartialEq, Debug, Clone)]
pub enum Type<'input> {
    // WHENEVER ADDING NEW PRIMITIVES HERE, ADD THEM TO THE TYPE SCOPE IN
    // [`zrc_typeck::typeck::scope::TypeScope::new`].
    /// `i8`
    I8,
    /// `u8`
    U8,
    /// `i16`
    I16,
    /// `u16`
    U16,
    /// `i32`
    I32,
    /// `u32`
    U32,
    /// `i64`
    I64,
    /// `u64`
    U64,
    /// `usize`
    Usize,
    /// `isize`
    Isize,
    /// `bool`
    Bool, /* TODO: need an "any Int" type that implicitly casts to all int types but becomes
           * i32 when assigned to a value */
    /// `*T`
    Ptr(Box<Type<'input>>),
    /// `fn(A, B) -> T`
    Fn(Fn<'input>),
    /// Struct type literals. Ordered by declaration order.
    Struct(IndexMap<&'input str, Type<'input>>),
    /// Union type literals. Ordered by declaration order.
    Union(IndexMap<&'input str, Type<'input>>),
    /// Opaque type placeholder used during type resolution for self-referential
    /// types. This is a temporary type that should be replaced with a void
    /// pointer (`*struct{}`) after the type definition is fully resolved.
    /// Opaque types should never appear in final TAST output or code
    /// generation.
    Opaque(&'input str),
}

impl Display for Type<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::I8 => write!(f, "i8"),
            Self::U8 => write!(f, "u8"),
            Self::I16 => write!(f, "i16"),
            Self::U16 => write!(f, "u16"),
            Self::I32 => write!(f, "i32"),
            Self::U32 => write!(f, "u32"),
            Self::I64 => write!(f, "i64"),
            Self::U64 => write!(f, "u64"),
            Self::Usize => write!(f, "usize"),
            Self::Isize => write!(f, "isize"),
            Self::Bool => write!(f, "bool"),
            Self::Ptr(pointee_ty) => write!(f, "*{pointee_ty}"),
            Self::Fn(fn_data) => write!(f, "{fn_data}"),
            Self::Struct(fields) if fields.is_empty() => write!(f, "struct {{}}"),
            Self::Struct(fields) => write!(
                f,
                "struct {{ {} }}",
                fields
                    .iter()
                    .map(|(key, ty)| format!("{key}: {ty}"))
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
            Self::Union(fields) if fields.is_empty() => write!(f, "union {{}}"),
            Self::Union(fields) => write!(
                f,
                "union {{ {} }}",
                fields
                    .iter()
                    .map(|(key, ty)| format!("{key}: {ty}"))
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
            Self::Opaque(name) => write!(f, "{name}"),
        }
    }
}

impl<'input> Type<'input> {
    /// Returns `true` if this is an integer type like [`Type::I8`].
    #[must_use]
    pub const fn is_integer(&self) -> bool {
        use Type::{I8, I16, I32, I64, Isize, U8, U16, U32, U64, Usize};
        matches!(
            self,
            I8 | U8 | I16 | U16 | I32 | U32 | I64 | U64 | Isize | Usize
        )
    }

    /// Returns `true` if this is a signed integer type like [`Type::I8`].
    #[must_use]
    pub const fn is_signed_integer(&self) -> bool {
        use Type::{I8, I16, I32, I64, Isize};
        matches!(self, I8 | I16 | I32 | I64 | Isize)
    }

    /// Returns `true` if this is an unsigned integer type like [`Type::U8`].
    #[must_use]
    pub const fn is_unsigned_integer(&self) -> bool {
        use Type::{U8, U16, U32, U64, Usize};
        matches!(self, U8 | U16 | U32 | U64 | Usize)
    }

    /// Try to get the value we point at, or None if not a pointer.
    #[must_use]
    #[allow(clippy::wildcard_enum_match_arm)]
    pub fn into_pointee(self) -> Option<Self> {
        match self {
            Type::Ptr(x) => Some(*x),
            _ => None,
        }
    }

    /// Try to access the struct's [`IndexMap`] if we are a struct
    #[must_use]
    #[allow(clippy::wildcard_enum_match_arm)]
    pub fn into_struct_contents(self) -> Option<IndexMap<&'input str, Self>> {
        match self {
            Type::Struct(x) => Some(x),
            _ => None,
        }
    }

    /// Try to access the union's [`IndexMap`] if we are a union
    #[must_use]
    #[allow(clippy::wildcard_enum_match_arm)]
    pub fn into_union_contents(self) -> Option<IndexMap<&'input str, Self>> {
        match self {
            Type::Union(x) => Some(x),
            _ => None,
        }
    }

    /// Get the unit type
    #[must_use]
    pub fn unit() -> Self {
        Type::Struct(IndexMap::new())
    }

    /// Check if this type can be implicitly cast to the target type.
    /// Currently only supports `*T` -> `*struct{}` (void pointer downcast).
    ///
    /// # Examples
    ///
    /// This allows any pointer type to implicitly downcast to void pointer:
    /// ```zirco
    /// fn takes_void_ptr(ptr: *struct{}) -> struct{} {
    ///     // ...
    /// }
    ///
    /// // All of these work without explicit casts:
    /// takes_void_ptr(&x);        // where x: i32
    /// takes_void_ptr(&y);        // where y: SomeStruct
    /// takes_void_ptr(&z);        // where z: bool
    /// ```
    #[must_use]
    pub fn can_implicitly_cast_to(&self, target: &Self) -> bool {
        // Allow any pointer type to implicitly cast to void pointer (*struct{})
        if let (Type::Ptr(_from_pointee), Type::Ptr(to_pointee)) = (self, target)
            && let Type::Struct(fields) = to_pointee.as_ref()
            && fields.is_empty()
        {
            // Target is *struct{}, allow implicit cast from any *T
            return true;
        }
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_void_ptr_implicit_cast() {
        // Create a void pointer type (*struct{})
        let void_ptr = Type::Ptr(Box::new(Type::Struct(IndexMap::new())));

        // Create various pointer types
        let i32_ptr = Type::Ptr(Box::new(Type::I32));
        let bool_ptr = Type::Ptr(Box::new(Type::Bool));
        let struct_ptr = Type::Ptr(Box::new(Type::Struct(IndexMap::from([("x", Type::I8)]))));

        // All should be able to implicitly cast to void pointer
        assert!(i32_ptr.can_implicitly_cast_to(&void_ptr));
        assert!(bool_ptr.can_implicitly_cast_to(&void_ptr));
        assert!(struct_ptr.can_implicitly_cast_to(&void_ptr));

        // Void pointer itself should also work
        assert!(void_ptr.can_implicitly_cast_to(&void_ptr));

        // Non-pointer types should not implicitly cast to void pointer
        assert!(!Type::I32.can_implicitly_cast_to(&void_ptr));

        // Void pointer should not implicitly cast to specific pointer
        assert!(!void_ptr.can_implicitly_cast_to(&i32_ptr));

        // Specific pointers should not implicitly cast to other specific pointers
        assert!(!i32_ptr.can_implicitly_cast_to(&bool_ptr));
    }
}
