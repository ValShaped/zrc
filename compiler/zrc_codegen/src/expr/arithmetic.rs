//! code generation for arithmetic expressions

use inkwell::{
    builder::BuilderError,
    values::{BasicValue, BasicValueEnum, IntValue},
};
use zrc_typeck::tast::{
    expr::{Arithmetic, BinaryBitwise, TypedExpr},
    ty::Type,
};

use crate::{
    bb::{BasicBlockAnd, BasicBlockExt},
    ctx::BlockCtx,
    expr::{CgExprArgs, cg_expr},
    ty::llvm_basic_type,
    unpack,
};

/// Build the required instruction for a [`BinaryBitwise`] operation
pub fn build_binary_bitwise<'ctx>(
    cg: BlockCtx<'ctx, '_, '_>,
    op: BinaryBitwise,
    lhs: IntValue<'ctx>,
    rhs: IntValue<'ctx>,
    result_is_signed: bool,
) -> Result<IntValue<'ctx>, BuilderError> {
    match op {
        BinaryBitwise::And => cg.builder.build_and(lhs, rhs, "and"),
        BinaryBitwise::Or => cg.builder.build_or(lhs, rhs, "or"),
        BinaryBitwise::Xor => cg.builder.build_xor(lhs, rhs, "xor"),
        BinaryBitwise::Shl => cg.builder.build_left_shift(lhs, rhs, "shl"),
        BinaryBitwise::Shr => cg
            .builder
            .build_right_shift(lhs, rhs, result_is_signed, "shr"),
    }
}
/// Build the required instruction for an [`Arithmetic`] operation
pub fn build_arithmetic<'ctx>(
    cg: BlockCtx<'ctx, '_, '_>,
    op: Arithmetic,
    lhs: IntValue<'ctx>,
    rhs: IntValue<'ctx>,
    result_is_signed: bool,
) -> Result<IntValue<'ctx>, BuilderError> {
    match op {
        Arithmetic::Addition => cg.builder.build_int_add(lhs, rhs, "add"),
        Arithmetic::Subtraction => cg.builder.build_int_sub(lhs, rhs, "sub"),
        Arithmetic::Multiplication => cg.builder.build_int_mul(lhs, rhs, "mul"),

        Arithmetic::Division if result_is_signed => {
            cg.builder.build_int_signed_div(lhs, rhs, "div")
        }
        Arithmetic::Division => cg.builder.build_int_unsigned_div(lhs, rhs, "div"),

        Arithmetic::Modulo if result_is_signed => cg.builder.build_int_signed_rem(lhs, rhs, "rem"),
        Arithmetic::Modulo => cg.builder.build_int_unsigned_rem(lhs, rhs, "rem"),
    }
}

/// Code generate a binary bitwise operation
pub fn cg_binary_bitwise<'ctx, 'input>(
    CgExprArgs {
        cg,
        mut bb,
        inferred_type,
        ..
    }: CgExprArgs<'ctx, 'input, '_>,
    op: BinaryBitwise,
    lhs: Box<TypedExpr<'input>>,
    rhs: Box<TypedExpr<'input>>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let lhs = unpack!(bb = cg_expr(cg, bb, *lhs));
    let rhs = unpack!(bb = cg_expr(cg, bb, *rhs));

    let reg = build_binary_bitwise(
        cg,
        op,
        lhs.into_int_value(),
        rhs.into_int_value(),
        inferred_type.is_signed_integer(),
    )
    .expect("binary bitwise operation should have compiled successfully");

    bb.and(reg.as_basic_value_enum())
}

/// Code generate an arithmetic operation
pub fn cg_arithmetic<'ctx, 'input>(
    CgExprArgs {
        cg,
        mut bb,
        inferred_type,
        ..
    }: CgExprArgs<'ctx, 'input, '_>,
    op: Arithmetic,
    lhs: Box<TypedExpr<'input>>,
    rhs: Box<TypedExpr<'input>>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let lhs_ty = lhs.inferred_type.clone();
    let lhs = unpack!(bb = cg_expr(cg, bb, *lhs));
    let rhs = unpack!(bb = cg_expr(cg, bb, *rhs));

    if let Type::Ptr(pointee) = lhs_ty {
        // Most languages make incrementing a pointer increase the address by the size
        // of the pointee type, hence our use of `gep`.
        // REVIEW: Is this the approach we want to take?
        #[allow(clippy::wildcard_enum_match_arm)]
        let reg = match op {
            // SAFETY: This can segfault if indices are used incorrectly
            // This is only used for pointer arithmetic, so the indices should be correct
            Arithmetic::Addition => unsafe {
                cg.builder.build_gep(
                    llvm_basic_type(&cg, &pointee).0,
                    lhs.into_pointer_value(),
                    &[rhs.into_int_value()],
                    "ptr_add",
                )
            },
            Arithmetic::Subtraction => {
                let rhs = cg
                    .builder
                    .build_int_neg(rhs.into_int_value(), "neg")
                    .expect("negation should have compiled successfully")
                    .as_basic_value_enum();

                // SAFETY: This can segfault if indices are used incorrectly
                // This is only used for pointer arithmetic, so the indices should be
                // correct
                unsafe {
                    cg.builder.build_gep(
                        llvm_basic_type(&cg, &pointee).0,
                        lhs.into_pointer_value(),
                        &[rhs.into_int_value()],
                        "ptr_sub",
                    )
                }
            }
            _ => panic!("invalid pointer arithmetic operation"),
        }
        .expect("pointer arithmetic should have compiled successfully");

        bb.and(reg.as_basic_value_enum())
    } else {
        let reg = build_arithmetic(
            cg,
            op,
            lhs.into_int_value(),
            rhs.into_int_value(),
            inferred_type.is_signed_integer(),
        )
        .expect("arithmetic operation should have compiled successfully");

        bb.and(reg.as_basic_value_enum())
    }
}

/// Code generate a unary bitwise NOT operation
pub fn cg_unary_bitwise_not<'ctx, 'input>(
    CgExprArgs { cg, mut bb, .. }: CgExprArgs<'ctx, 'input, '_>,
    x: Box<TypedExpr<'input>>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let value = unpack!(bb = cg_expr(cg, bb, *x));

    let reg = cg
        .builder
        .build_not(value.into_int_value(), "not")
        .expect("not should have compiled successfully");

    bb.and(reg.as_basic_value_enum())
}

/// Code generate a unary minus operation
pub fn cg_unary_minus<'ctx, 'input>(
    CgExprArgs { cg, mut bb, .. }: CgExprArgs<'ctx, 'input, '_>,
    x: Box<TypedExpr<'input>>,
) -> BasicBlockAnd<'ctx, BasicValueEnum<'ctx>> {
    let value = unpack!(bb = cg_expr(cg, bb, *x));

    let reg = cg
        .builder
        .build_int_neg(value.into_int_value(), "neg")
        .expect("negation should have compiled successfully");

    bb.and(reg.as_basic_value_enum())
}

#[cfg(test)]
mod tests {
    // Please read the "Common patterns in tests" section of crate::test_utils for
    // more information on how code generator tests are structured.

    use indoc::indoc;

    use crate::cg_snapshot_test;

    #[test]
    fn pointer_arithmetic_generates_proper_gep() {
        cg_snapshot_test!(indoc! {r#"
                fn test() {
                    let x: *i32;

                    // TEST: should create a GEP that is the same as "x[4]"
                    let y = x + 4 as usize;
                    // TEST: and the same, with -4:
                    let z = x - 4 as usize;
                }
            "#});
    }

    #[test]
    fn arithmetic_operators_generate() {
        cg_snapshot_test!(indoc! {"
                fn get_int() -> i32;
                fn get_uint() -> u32;

                fn test() {
                    let sx = get_int();
                    let sy = get_int();
                    let ux = get_uint();
                    let uy = get_uint();

                    // TEST: should create an `add i32` instruction
                    let s_add = sx + sy;
                    let u_add = ux + uy;

                    // TEST: should create a `sub i32` instruction
                    let s_sub = sx - sy;
                    let u_sub = ux - uy;

                    // TEST: should create a `mul i32` instruction
                    let s_mul = sx * sy;
                    let u_mul = ux * uy;

                    // TEST: should create a `sdiv i32` instruction
                    let s_div = sx / sy;
                    // TEST: should create a `udiv i32` instruction
                    let u_div = ux / uy;

                    // TEST: should create a `srem i32` instruction
                    let s_rem = sx % sy;
                    // TEST: should create a `urem i32` instruction
                    let u_rem = ux % uy;
                }
            "});
    }

    #[test]
    fn bitwise_operators_generate() {
        cg_snapshot_test!(indoc! {"
                fn get_int() -> i32;
                fn get_uint() -> u32;

                fn test() {
                    let x = get_int();
                    let y = get_int();
                    let u = get_uint();

                    // TEST: should create a `not i32` instruction
                    let not = ~x;

                    // TEST: should create a `and i32` instruction
                    let and = x & y;

                    // TEST: should create a `or i32` instruction
                    let or = x | y;

                    // TEST: should create a `xor i32` instruction
                    let xor = x ^ y;

                    // TEST: should create a `shl i32` instruction
                    let shl = x << u;

                    // TEST: should create a `lshr i32` instruction
                    let lshr = u >> u;

                    // TEST: should create a `ashr i32` instruction (as the lhs is signed)
                    let ashr = x >> u;
                }
            "});
    }
}
