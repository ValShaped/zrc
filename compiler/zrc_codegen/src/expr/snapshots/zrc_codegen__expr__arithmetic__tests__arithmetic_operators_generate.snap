---
source: compiler/zrc_codegen/src/expr/arithmetic.rs
description: "fn get_int() -> i32;\nfn get_uint() -> u32;\n\nfn test() {\n    let sx = get_int();\n    let sy = get_int();\n    let ux = get_uint();\n    let uy = get_uint();\n\n    // TEST: should create an `add i32` instruction\n    let s_add = sx + sy;\n    let u_add = ux + uy;\n\n    // TEST: should create a `sub i32` instruction\n    let s_sub = sx - sy;\n    let u_sub = ux - uy;\n\n    // TEST: should create a `mul i32` instruction\n    let s_mul = sx * sy;\n    let u_mul = ux * uy;\n\n    // TEST: should create a `sdiv i32` instruction\n    let s_div = sx / sy;\n    // TEST: should create a `udiv i32` instruction\n    let u_div = ux / uy;\n\n    // TEST: should create a `srem i32` instruction\n    let s_rem = sx % sy;\n    // TEST: should create a `urem i32` instruction\n    let u_rem = ux % uy;\n}\n"
expression: resulting_ir
---
; ModuleID = 'test.zr'
source_filename = "test.zr"

declare i32 @get_int()

declare i32 @get_uint()

define {} @test() !dbg !3 {
entry:
  %let_u_rem = alloca i32, align 4
  %let_s_rem = alloca i32, align 4
  %let_u_div = alloca i32, align 4
  %let_s_div = alloca i32, align 4
  %let_u_mul = alloca i32, align 4
  %let_s_mul = alloca i32, align 4
  %let_u_sub = alloca i32, align 4
  %let_s_sub = alloca i32, align 4
  %let_u_add = alloca i32, align 4
  %let_s_add = alloca i32, align 4
  %let_uy = alloca i32, align 4
  %let_ux = alloca i32, align 4
  %let_sy = alloca i32, align 4
  %let_sx = alloca i32, align 4
  %call = call i32 @get_int(), !dbg !8
  store i32 %call, ptr %let_sx, align 4, !dbg !11
  %call1 = call i32 @get_int(), !dbg !12
  store i32 %call1, ptr %let_sy, align 4, !dbg !13
  %call2 = call i32 @get_uint(), !dbg !14
  store i32 %call2, ptr %let_ux, align 4, !dbg !15
  %call3 = call i32 @get_uint(), !dbg !16
  store i32 %call3, ptr %let_uy, align 4, !dbg !17
  %load = load i32, ptr %let_sx, align 4, !dbg !18
  %load4 = load i32, ptr %let_sy, align 4, !dbg !19
  %add = add i32 %load, %load4, !dbg !19
  store i32 %add, ptr %let_s_add, align 4, !dbg !20
  %load5 = load i32, ptr %let_ux, align 4, !dbg !21
  %load6 = load i32, ptr %let_uy, align 4, !dbg !22
  %add7 = add i32 %load5, %load6, !dbg !22
  store i32 %add7, ptr %let_u_add, align 4, !dbg !23
  %load8 = load i32, ptr %let_sx, align 4, !dbg !24
  %load9 = load i32, ptr %let_sy, align 4, !dbg !25
  %sub = sub i32 %load8, %load9, !dbg !25
  store i32 %sub, ptr %let_s_sub, align 4, !dbg !26
  %load10 = load i32, ptr %let_ux, align 4, !dbg !27
  %load11 = load i32, ptr %let_uy, align 4, !dbg !28
  %sub12 = sub i32 %load10, %load11, !dbg !28
  store i32 %sub12, ptr %let_u_sub, align 4, !dbg !29
  %load13 = load i32, ptr %let_sx, align 4, !dbg !30
  %load14 = load i32, ptr %let_sy, align 4, !dbg !31
  %mul = mul i32 %load13, %load14, !dbg !31
  store i32 %mul, ptr %let_s_mul, align 4, !dbg !32
  %load15 = load i32, ptr %let_ux, align 4, !dbg !33
  %load16 = load i32, ptr %let_uy, align 4, !dbg !34
  %mul17 = mul i32 %load15, %load16, !dbg !34
  store i32 %mul17, ptr %let_u_mul, align 4, !dbg !35
  %load18 = load i32, ptr %let_sx, align 4, !dbg !36
  %load19 = load i32, ptr %let_sy, align 4, !dbg !37
  %div = sdiv i32 %load18, %load19, !dbg !37
  store i32 %div, ptr %let_s_div, align 4, !dbg !38
  %load20 = load i32, ptr %let_ux, align 4, !dbg !39
  %load21 = load i32, ptr %let_uy, align 4, !dbg !40
  %div22 = udiv i32 %load20, %load21, !dbg !40
  store i32 %div22, ptr %let_u_div, align 4, !dbg !41
  %load23 = load i32, ptr %let_sx, align 4, !dbg !42
  %load24 = load i32, ptr %let_sy, align 4, !dbg !43
  %rem = srem i32 %load23, %load24, !dbg !43
  store i32 %rem, ptr %let_s_rem, align 4, !dbg !44
  %load25 = load i32, ptr %let_ux, align 4, !dbg !45
  %load26 = load i32, ptr %let_uy, align 4, !dbg !46
  %rem27 = urem i32 %load25, %load26, !dbg !46
  store i32 %rem27, ptr %let_u_rem, align 4, !dbg !47
  ret {} zeroinitializer, !dbg !48
}

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "zrc test runner", isOptimized: false, flags: "zrc --fake-args", runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false)
!2 = !DIFile(filename: "test.zr", directory: "/fake/path")
!3 = distinct !DISubprogram(name: "test", linkageName: "test", scope: null, file: !2, line: 4, type: !4, scopeLine: 4, spFlags: DISPFlagDefinition, unit: !1)
!4 = !DISubroutineType(types: !5)
!5 = !{!6}
!6 = !DICompositeType(tag: DW_TAG_structure_type, name: "struct {}", scope: !2, file: !2, elements: !7)
!7 = !{}
!8 = !DILocation(line: 5, column: 14, scope: !9)
!9 = distinct !DILexicalBlock(scope: !10, file: !2, line: 4, column: 11)
!10 = distinct !DILexicalBlock(scope: !3, file: !2, line: 4, column: 11)
!11 = !DILocation(line: 5, column: 9, scope: !9)
!12 = !DILocation(line: 6, column: 14, scope: !9)
!13 = !DILocation(line: 6, column: 9, scope: !9)
!14 = !DILocation(line: 7, column: 14, scope: !9)
!15 = !DILocation(line: 7, column: 9, scope: !9)
!16 = !DILocation(line: 8, column: 14, scope: !9)
!17 = !DILocation(line: 8, column: 9, scope: !9)
!18 = !DILocation(line: 11, column: 17, scope: !9)
!19 = !DILocation(line: 11, column: 22, scope: !9)
!20 = !DILocation(line: 11, column: 9, scope: !9)
!21 = !DILocation(line: 12, column: 17, scope: !9)
!22 = !DILocation(line: 12, column: 22, scope: !9)
!23 = !DILocation(line: 12, column: 9, scope: !9)
!24 = !DILocation(line: 15, column: 17, scope: !9)
!25 = !DILocation(line: 15, column: 22, scope: !9)
!26 = !DILocation(line: 15, column: 9, scope: !9)
!27 = !DILocation(line: 16, column: 17, scope: !9)
!28 = !DILocation(line: 16, column: 22, scope: !9)
!29 = !DILocation(line: 16, column: 9, scope: !9)
!30 = !DILocation(line: 19, column: 17, scope: !9)
!31 = !DILocation(line: 19, column: 22, scope: !9)
!32 = !DILocation(line: 19, column: 9, scope: !9)
!33 = !DILocation(line: 20, column: 17, scope: !9)
!34 = !DILocation(line: 20, column: 22, scope: !9)
!35 = !DILocation(line: 20, column: 9, scope: !9)
!36 = !DILocation(line: 23, column: 17, scope: !9)
!37 = !DILocation(line: 23, column: 22, scope: !9)
!38 = !DILocation(line: 23, column: 9, scope: !9)
!39 = !DILocation(line: 25, column: 17, scope: !9)
!40 = !DILocation(line: 25, column: 22, scope: !9)
!41 = !DILocation(line: 25, column: 9, scope: !9)
!42 = !DILocation(line: 28, column: 17, scope: !9)
!43 = !DILocation(line: 28, column: 22, scope: !9)
!44 = !DILocation(line: 28, column: 9, scope: !9)
!45 = !DILocation(line: 30, column: 17, scope: !9)
!46 = !DILocation(line: 30, column: 22, scope: !9)
!47 = !DILocation(line: 30, column: 9, scope: !9)
!48 = !DILocation(line: 31, column: 1, scope: !9)
